lib/logstash/outputs/elasticsearch/http_client/pool.rb

```diff
@@ -282,19 +283,9 @@ def healthcheck!
               @logger.warn("Detected a node with a higher major version than previously observed. This could be the result of an elasticsearch cluster upgrade.", :previous_major => @maximum_seen_major_version, :new_major => major, :node_url => url.sanitized.to_s)
               set_new_major_version(major)
             end
-            if oss? || valid_es_license?(url)
-              meta[:state] = :alive
-            else
-              # As this version is to be shipped with Logstash 7.x we won't mark the connection as unlicensed
-              #
-              #  logger.error("Cannot connect to the Elasticsearch cluster configured in the Elasticsearch output. Logstash requires the default distribution of Elasticsearch. Please update to the default distribution of Elasticsearch for full access to all free features, or switch to the OSS distribution of Logstash.", :url => url.sanitized.to_s)
-              #  meta[:state] = :unlicensed
-              #
-              # Instead we'll log a deprecation warning and mark it as alive:
-              #
-              log_license_deprecation_warn(url)
-              meta[:state] = :alive
-            end
+
+            alive = @license_checker.appropriate_license?(self, url)
+            meta[:state] = alive ? :alive : :unlicensed
```

Comment:
so would `:dead` be the right state for this then?  
The valid states and their meanings should ideally be encoded somewhere as constants or such. No sure we want to introduce that refactor here but we should clean that up at some point.

Reply:
So looking more closely at the code the 2 places where meta is really used is:

- https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/efc1c069d04444f77db1549d95e4f06131497f11/lib/logstash/outputs/elasticsearch/http_client/pool.rb#L267 

`healthcheck!` cycles through all `@url_info` with a meta other than `:alive` to see if it can be reused, so here any value, like `:unlicensed` is ok, a health check will be performed periodically on these.

- https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/efc1c069d04444f77db1549d95e4f06131497f11/lib/logstash/outputs/elasticsearch/http_client/pool.rb#L441-L451

`get_connection` will pick a connection if it is not `:dead`. Here I am not sure if the logic should not be to only consider the `:alive` ones instead of the not `:dead` ones if we introduce another state like `:unlicensed`. The question is: would is make sense for `get_connection` to return a `:unlicensed` connection? I wouldn't think so. 

Currently a `:dead` connection is set upon `HostUnreachableError` in https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/efc1c069d04444f77db1549d95e4f06131497f11/lib/logstash/outputs/elasticsearch/http_client/pool.rb#L402-L405.

So with this, it seems that we could also simply set the `meta[:state] = :dead` instead of a new `:unlicensed` state and we would not have to change anything else in the logic. The above 2 places where meta is looked at would work; `healthcheck!`  would perform a check on a `:dead` url and `get_connection`  would not pick it up if it was maked as `:dead`. 

LMKWYT.